96.md

```diff
@@ -0,0 +1,192 @@
+NIP-96
+======
+
+HTTP File Storage Integration
+-----------------------------
+
+`draft` `optional` `author:arthurfranca` `author:Semisol` `author:staab`
+
+## Introduction
+
+This NIP defines a REST API for HTTP file storage servers intended to be used in conjunction with the nostr network.
+The API will enable nostr users to upload files and later reference them by url on nostr notes.
+
+The spec DOES NOT use regular nostr events through websockets for
+storing, requesting nor retrieving data because, for simplicity, the server
+will not have to learn anything about nostr relays.
+
+## Server Adaptation
+
+File storage servers wishing to be accessible by nostr users should opt-in by making available
+an https route at `/.well-known/nostr.json` with `nip96.routes` field in the following form:
+
+```js
+{
+  "nip96": {
+    // required
+    "routes": {
+      "upload": "https://upload.domain/upload-path",
+      "download": "https://download.domain/download-path",
+      "deletion": "https://deletion.domain/deletion-path"
+    },
+    // optional
+    "supported_nips": [],
+    // optional
+    "tos_url": "https://my.landing.page/terms-of-service",
+    // optional
+    "plans": {
+      // "free" is the only standardized plan key and
+      // clients may use its presence to learn if server offers free storage
+      "free": {
+        "name": "Free Tier",
+        "url": "https://...", // payment etc
+        "max_byte_size": 10485760,
+        // range in days / 0 for no expiration
+        // [7, 0] means it may vary from 7 days to unlimited persistence,
+        // [0, 0] means it has no expiration
+        // early expiration may be due to low traffic or any other factor
+        "file_expiration": [14, 90],
+        "image_transformations": ["resizing"],
+        "cost": {
+          // it considers lowest currency unit
+          // "100" USD equals 100 cents or 1 dollar
+          // "1000" BTC equals 1000 milisats or 1 satoshi
+          "price": "0", // string
+          "currency": "USD",
+          // "per-upload|one-time|day|week|month|year..."
+          "frequency": ""
+        }
+      }
+    }
+  }
+}
+```
+
+Note that if using a CDN, the `server` should issue a 302 redirect from
+the `nip96.routes.download` route to the CDN url instead
+of assigning the CDN directly as the `nip96.routes.download` route.
+This is because some NIP-96 integrations
+may require NIP-98 authentication for downloads,
+which the CDN won't be able to support.
```

Comment:
Upload/download/deletion at same root_url would surely make things simpler.

The text you quoted is warning for the case of some future nip integrations (like NIP-60)
that requires NIP-98 authentication for downloads.

But for other cases such as free downloads, unfortunatelly, some hosts may want to point directly to a CDN
without the redirect to prevent files from going offline when the API is down.

Hey, @bndw @michaelhall923 and @v0l can you confirm if the api-to-cdn redirect would or not be acceptable for downloads?

Reply:
Now there is api_url (that may include any path) and optional download_url.